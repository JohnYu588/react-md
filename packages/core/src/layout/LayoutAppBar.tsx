import { forwardRef } from "react";
import { AppBar } from "../app-bar";
import { LayoutAppBarTitle } from "./LayoutAppBarTitle";
import { LayoutNavToggle } from "./LayoutNavToggle";
import { useLayoutConfig } from "./LayoutProvider";
import type { LayoutAppBarProps } from "./types";

/**
 * This is the default implementation for an `AppBar` within the `Layout` that
 * will conditionally render the default `LayoutNavToggle` button and
 * `AppBarTitle` depending on specific props that were provided.
 */
export const LayoutAppBar = forwardRef<HTMLDivElement, LayoutAppBarProps>(
  function LayoutAppBar(props, ref) {
    const {
      children,
      fixed = true,
      navToggle: propNavToggle,
      navToggleProps,
      customTitle,
      title: titleChildren,
      titleProps,
      scrollbarOffset = true,
      ...remaining
    } = props;
    const { baseId, layout } = useLayoutConfig();

    let nav = propNavToggle;
    if (typeof nav === "undefined") {
      // set the key to the current layout since we want the button to re-mount
      // on layout changes so the transition does not occur
      nav = <LayoutNavToggle key={layout} {...navToggleProps} />;
    }

    let title = customTitle;
    if (typeof title === "undefined") {
      title = (
        <LayoutAppBarTitle keyline={nav ? "nav" : undefined} {...titleProps}>
          {titleChildren}
        </LayoutAppBarTitle>
      );
    }

    return (
      <AppBar
        id={`${baseId}-header`}
        {...remaining}
        ref={ref}
        fixed={fixed}
        scrollbarOffset={scrollbarOffset}
      >
        {nav}
        {title}
        {children}
      </AppBar>
    );
  }
);
