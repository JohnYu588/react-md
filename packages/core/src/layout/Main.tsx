"use client";
import type { ElementType, HTMLAttributes, ReactNode, Ref } from "react";
import { forwardRef } from "react";
import { useEnsuredId } from "../useEnsuredId.js";
import type { MainClassNameOptions } from "./mainStyles.js";
import { main } from "./mainStyles.js";
import { useMainTabIndex } from "./useMainTabIndex.js";

/**
 * @remarks \@since 6.0.0
 */
export type CustomMainElement = ElementType<
  HTMLAttributes<HTMLElement> & {
    ref: Ref<HTMLElement>;
    className?: string;
    tabIndex?: number;
  }
>;

/**
 * @remarks \@since 6.0.0
 */
export interface MainProps
  extends HTMLAttributes<HTMLElement>,
    MainClassNameOptions {
  /**
   * @defaultValue `"main"`
   */
  as?: CustomMainElement;
  children: ReactNode;
}

/**
 * **Client Component**
 *
 * This component is really only used to dynamically set the `tabIndex` to `-1`
 * while using a keyboard for the `SkipToMainContent` component's focus behavior
 * to work correctly. If you don't need that functionality, use the {@link main}
 * style utility instead.
 *
 * @example
 * Styles Only
 * ```tsx
 * import { main as mainStyles } from "@react-md/core":
 *
 * function MyCustomMainElement({ children }) {
 *   return (
 *     <main
 *       className={mainStyles({
 *         navOffset: true,
 *         appBarOffset: true,
 *       })}
 *     >
 *       {children}
 *     </main>
 *   );
 * }
 * ```
 *
 * @remarks \@since 6.0.0 Renamed from `LayoutMain` removed a lot of
 * functionality to keep this component simple.
 */
export const Main = forwardRef<HTMLElement, MainProps>(
  function Main(props, ref) {
    const {
      as: Component = "main",
      id: propId,
      className,
      children,
      tabIndex: propTabIndex,
      navOffset,
      appBarOffset,
      ...remaining
    } = props;
    const id = useEnsuredId(propId, "main");
    const tabIndex = useMainTabIndex(propTabIndex);

    return (
      <Component
        {...remaining}
        id={id}
        ref={ref}
        className={main({ navOffset, appBarOffset, className })}
        tabIndex={tabIndex}
      >
        {children}
      </Component>
    );
  }
);
