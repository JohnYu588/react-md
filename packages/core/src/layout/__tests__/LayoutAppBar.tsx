import { describe, expect, it, jest } from "@jest/globals";
import { createRef, type ReactElement, type Ref } from "react";
import {
  act,
  cleanupResizeObserverAfterEach,
  rmdRender,
  screen,
  setupResizeObserverMock,
} from "../../test-utils/index.js";
import { LayoutAppBar, type LayoutAppBarProps } from "../LayoutAppBar.js";
import { Main } from "../Main.js";

const getVar = () =>
  window
    .getComputedStyle(document.documentElement)
    .getPropertyValue("--rmd-layout-header-height");

function Test(
  props: LayoutAppBarProps & { nodeRef?: Ref<HTMLDivElement> }
): ReactElement {
  const { nodeRef, ...remaining } = props;
  return (
    <>
      <LayoutAppBar {...remaining} ref={nodeRef} />
      <Main>Hello, world!</Main>
    </>
  );
}

cleanupResizeObserverAfterEach();

describe("LayoutAppBar", () => {
  it("should apply the correct styling, HTMLAttributes, and allow a ref", () => {
    const ref = createRef<HTMLDivElement>();
    const props = {
      nodeRef: ref,
      children: "Content",
    } as const;

    const { rerender } = rmdRender(<Test {...props} />);

    const appBar = screen.getByRole("banner");
    expect(ref.current).toBeInstanceOf(HTMLElement);
    expect(ref.current).toBe(appBar);
    expect(appBar).toMatchSnapshot();

    rerender(
      <Test
        {...props}
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(appBar).toMatchSnapshot();

    rerender(<Test {...props} position="sticky" />);
    expect(appBar).toMatchSnapshot();

    rerender(<Test {...props} as="div" data-testid="div" />);
    const divAppBar = screen.getByTestId("div");
    expect(ref.current).toBeInstanceOf(HTMLDivElement);
    expect(ref.current).toBe(divAppBar);
    expect(divAppBar).toMatchSnapshot();
  });

  it("should automatically apply the app bar height as css variables", () => {
    const observer = setupResizeObserverMock();
    rmdRender(<Test />);

    expect(getVar()).toBe("0px");
    const appBar = screen.getByRole("banner");
    act(() => {
      observer.resizeElement(appBar, {
        height: 96,
      });
    });
    expect(getVar()).toBe("96px");
  });

  it("should throw an if there is no main element since it renders the SkipToMainContent component", () => {
    const error = jest.spyOn(console, "error").mockImplementation(() => {});
    expect(() => rmdRender(<LayoutAppBar />)).toThrow();
    error.mockRestore();
  });

  it("should allow props to be passed to the SkipToMainContent component", () => {
    rmdRender(
      <Test skipProps={{ children: "Main", className: "custom-link-class" }} />
    );

    const link = screen.getByRole("link");
    expect(link).toMatchSnapshot();
  });
});
