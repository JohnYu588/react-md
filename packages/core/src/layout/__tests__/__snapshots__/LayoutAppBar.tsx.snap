// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutAppBar should allow props to be passed to the SkipToMainContent component 1`] = `
<a
  class="rmd-link rmd-link rmd-link--skip rmd-link--skip-styled custom-link-class"
  href="#"
  id="skip-to-main"
>
  Main
</a>
`;

exports[`LayoutAppBar should apply the correct styling, HTMLAttributes, and allow a ref 1`] = `
<header
  class="rmd-app-bar rmd-app-bar--fixed rmd-app-bar--top rmd-app-bar--elevated rmd-app-bar--scrollbar-offset rmd-background-container rmd-primary-container rmd-full-width"
  id="layout-header-:r0:"
>
  <a
    class="rmd-link rmd-link rmd-link--skip rmd-link--skip-styled"
    href="#"
    id="skip-to-main"
  >
    Skip to main content
  </a>
  Content
</header>
`;

exports[`LayoutAppBar should apply the correct styling, HTMLAttributes, and allow a ref 2`] = `
<header
  class="rmd-app-bar rmd-app-bar--fixed rmd-app-bar--top rmd-app-bar--elevated rmd-app-bar--scrollbar-offset rmd-background-container rmd-primary-container rmd-full-width custom-class-name"
  id="layout-header-:r0:"
  style="color: white;"
>
  <a
    class="rmd-link rmd-link rmd-link--skip rmd-link--skip-styled"
    href="#"
    id="skip-to-main"
  >
    Skip to main content
  </a>
  Content
</header>
`;

exports[`LayoutAppBar should apply the correct styling, HTMLAttributes, and allow a ref 3`] = `
<header
  class="rmd-app-bar rmd-app-bar--fixed rmd-app-bar--sticky rmd-app-bar--top rmd-app-bar--elevated rmd-background-container rmd-primary-container rmd-full-width"
  id="layout-header-:r0:"
  style=""
>
  <a
    class="rmd-link rmd-link rmd-link--skip rmd-link--skip-styled"
    href="#"
    id="skip-to-main"
  >
    Skip to main content
  </a>
  Content
</header>
`;

exports[`LayoutAppBar should apply the correct styling, HTMLAttributes, and allow a ref 4`] = `
<div
  class="rmd-app-bar rmd-app-bar--fixed rmd-app-bar--top rmd-app-bar--elevated rmd-app-bar--scrollbar-offset rmd-background-container rmd-primary-container rmd-full-width"
  data-testid="div"
  id="layout-header-:r0:"
>
  <a
    class="rmd-link rmd-link rmd-link--skip rmd-link--skip-styled"
    href="#"
    id="skip-to-main"
  >
    Skip to main content
  </a>
  Content
</div>
`;
